### ISSUE TITLE
OPTIONS request returning 404 

### ISSUE THREAD

**Issue by Jrfidellis** [FoalTS/foal#259]
I created a simple API with a 'hand made' login (not following the login provided by foal).

The API is meant to be public, without SSR.
 
The problem is that when the browser calls the API (more specifically when it does a OPTIONS request) the API returns a 404 because there is not a response to OPTIONS configured.

I tried adding the following code:
```
 const app = createApp(AppController, {
    store: session => new (sqliteStoreFactory(session))({ db: 'db.sqlite3' })
  })
  app.options('/*', (req, res, next) => {
    res.header('Access-Control-Allow-Origin', '*')
    res.header('Access-Control-Allow-Methods', 'GET,PUT,POST,DELETE,OPTIONS')
    res.header('Access-Control-Allow-Headers', 'Content-Type, Authorization, Content-Length, X-Requested-With')
    res.send(200)
  })
```

But the code inside the `app.options` is never called. I believe that this is because the `app.options` is being set after the routes, and there isn't a way to set it before them.

Maybe if something like that (code below) was possible this issue would not happen? 
```
// inside index.ts
const app = express()
app.options('/*', (req, res, next) => {
   res.header('Access-Control-Allow-Origin', '*')
   res.header('Access-Control-Allow-Methods', 'GET,PUT,POST,DELETE,OPTIONS')
   res.header('Access-Control-Allow-Headers', 'Content-Type, Authorization, Content-Length, X-Requested-With')
   res.send(200)
})
```
Then i would pass my `app` to the createApp, in a optional parameter. If `app` ins't provided then foal calls the `const app = express()` inside the createApp function:
```
// inside index.ts
const app = createApp(AppController, {
    store: session => new (sqliteStoreFactory(session))({ db: 'db.sqlite3' })
  }, app)
```

Maybe this issue is related to #256?

**Comment by Jrfidellis** [comment]
PS: @LoicPoullain, i would like your guidance to solve this issue as quick as possible, because i am planning to use Foal in a Hackathon this week, and this issue would be a problem for that. Thanks!

**Comment by LoicPoullain** [comment]
> But the code inside the app.options is never called. I believe that this is because the app.options is being set after the routes

Correct.

> Maybe this issue is related to #256?

Yes, it is. 

The problem here is that you are requesting a URL which is different from the `Origin` (the domain or even just the port may differ). This is fine but we need to tell the browser to enable *Cross-Origin Request Sharing*, which is what you are trying to do. [This article](https://www.html5rocks.com/en/tutorials/cors/) explains well this phenomenon.

To do so, I just released the [version 0.6.0](https://github.com/FoalTS/foal/releases/tag/v0.6.0) of FoalTS.

The right approach is to use this pattern:

```typescript
@Hook(() => (ctx, services, response) => {
  // Every response of this controller and its sub-controllers will be added this header.
  response.setHeader('Access-Control-Allow-Origin', '*');
})
export class ApiController {

  subControllers = [
    // your sub-controllers
  ];

  @Options('*')
  options(ctx: Context) {
    const response = new HttpResponseOK();
    // You may need to allow other headers depending on what you need.
    response.setHeader('Access-Control-Allow-Headers', 'Content-Type');
    return response;
  }

  // Some other routes (ex: @Get('/users'), etc)

}
```

This is the preferred approach as it does not use express middlewares. In a more distant future, FoalTS will no longer need to use Express as sub-layer (ideal situation). However if you really get into trouble to make this work, you can use the feature that you suggested that I just implemented:

```typescript
import * as cors from 'cors';
import * as express from 'express';

// ...

  const expressInstance = express();
  expressInstance.use(cors());

  const app = createApp(AppController, {
    store: session => new (sqliteStoreFactory(session))({ db: 'db.sqlite3' })
  }, expressInstance);

```

Let me know if this is not clear. Also if you have any feedbacks on the framework during your hackathon, let me know, I'll take them all!

Note: If you want to disable the csrf protection, you can do it in `config/settings.json`.

**Comment by Jrfidellis** [comment]
Thank you @LoicPoullain! I liked your approach of adding the `@Options`.

> Note: If you want to disable the csrf protection, you can do it in config/settings.json.

This gave me 5 seconds of headache during my first tests, lol, then i tried to do a `Ctrl+Shift-F` on `csrf` and found it! Very intuitive.
